components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -0,0 +1,174 @@
+#include "KeyValuePairLogEvent.hpp"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <system_error>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <outcome/single-header/outcome.hpp>
+
+#include "../ir/EncodedTextAst.hpp"
+#include "../time_types.hpp"
+#include "SchemaTree.hpp"
+#include "SchemaTreeNode.hpp"
+#include "Value.hpp"
+
+using clp::ir::EightByteEncodedTextAst;
+using clp::ir::FourByteEncodedTextAst;
+using std::string;
+
+namespace clp::ffi {
+namespace {
+/**
+ * @param type
+ * @param value
+ * @return Whether the given schema tree node type matches the given value's type.
+ */
+[[nodiscard]] auto
+node_type_matches_value_type(SchemaTreeNode::Type type, Value const& value) -> bool;
+
+/**
+ * Validates whether the given node ID value pairs are valid in terms of the key-value pair IR spec.
+ * @param schema_tree
+ * @param node_id_value_pairs
+ * @return `std::nullopt` if the inputs are valid, or an error code indicating the failure:
+ * - std::errc::operation_not_permitted if a node ID doesn't represent a valid node in the
+ *   schema tree, or a non-leaf node ID is paired with a value.
+ * - std::errc::protocol_error if the schema tree node type doesn't match the value's type.
+ * - std::errc::protocol_not_supported if the same key appears more than once under a parent
+ *   node.
+ */
+[[nodiscard]] auto validate_node_id_value_pairs(
+        SchemaTree const& schema_tree,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> std::optional<std::errc>;
+
+/**
+ * @param schema_tree
+ * @param node_id
+ * @param node_id_value_pairs
+ * @return Whether the given node is a leaf node in the sub schema tree defined by
+ * `node_id_value_pairs`. A node is considered a leaf if none of its descendants appear in the
+ * `node_id_value_pairs`.
+ */
+[[nodiscard]] auto is_leaf_node(
```

Comment:
Disclaimer: I only have a rough understanding on how schema tree works.
This function is bit confusing for me. 
"A node is considered a leaf if none of its descendants appear in the `node_id_value_pairs`." <- This doesn't seem to be the normal definition for a leaf node. Is this expected?

In addition, "the sub schema tree defined by `node_id_value_pairs`." Is there any guarantee that nodes in the `node_id_value_pairs` will form a subtree? Just looking at the argument without thinking about how it is generated, it is possible node_id_value_pairs contains a list of unconnected nodes.

And lastly a highlevel comment is that I feel this function should belong to the Schema tree class.

Reply:
This is a validation method: if an object type node appears in the key value pairs, it must be a leaf node. The way we store key value pairs is to store the schema (all as leaf nodes), so the path from the root to the leaves are implicitly indicated. In the merged schema tree, usually an object type node is a non leaf node. But in a key value pairs' schema, this node can be a leaf node as long as its value is `{}` or `null`. Our goal is to validate that if such a node appears, all its descendants in the merged schema tree can't appear in the schema of the given key value pair log event. `it is possible node_id_value_pairs contains a list of unconnected nodes` isn't true since everything will be connected to the root if they are valid leaf nodes. This function is used to validate pairs using schema tree's information. I think we should split it from a schema tree's method.